### PR TITLE
Greatly improve sidebar when width < 700px

### LINKED ISSUE #36531: Crates sidebar menu in docs disappears when page width is less than 700px
This can be confusing; it makes it look like there aren't any docs available. Example:

@bruceadams was following along with [Chapter 2 in the new book](http://rust-lang.github.io/book/ch02-00-guessing-game-tutorial.html#Generating%20a%20Random%20Number), and we mention `cargo doc --open` after adding the regex crate. At the time he ran the command, his browser window was under 700px wide because he had his screen split to have the book and editor/terminal open at the same time, so the docs opened themselves in that narrow window. The menu with the crates in the sidebars is just totally gone at that width, and that combined with the fact that the guessing game doesn't have any docs that would appear on the index page makes the docs look broken and like there aren't any docs to browse :(

Screenshot of this scenario:

screen shot 2016-09-16 at 9 33 26 am

The problem is this CSS:

``` css
@media (max-width: 700px) {
   // ...
  .sidebar .block {
    display: none;
  }
  // ...
}
```

It'd be awesome to have at least a hamburger menu or something to be able to get to the menu of crates even at narrower widths.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #36531.

r? @QuietMisdreavus 

A few screenshots:

screen shot 2017-12-06 at 00 41 36
screen shot 2017-12-06 at 00 41 40

### CODE REVIEW

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -3625,6 +3626,8 @@ impl<'a> fmt::Display for Sidebar<'a> {
             write!(fmt, "<script defer src=\"{path}sidebar-items.js\"></script>",
                    path = relpath)?;
         }
+        // Closes sidebar-elems div.
+        write!(fmt, "</div>")?;
```
Comment (kennytm) [inline src/librustdoc/html/render.rs]: Is this `</div>` placed correctly? It is at different indentation level than line 3578. htmldocck on CI currently complains all files are invalid HTML.

```
[01:05:10] stderr:
[01:05:10] ------------------------------------------
[01:05:10] Traceback (most recent call last):
[01:05:10]   File "/checkout/src/etc/htmldocck.py", line 455, in <module>
[01:05:10]     check(sys.argv[1], get_commands(sys.argv[2]))
[01:05:10]   File "/checkout/src/etc/htmldocck.py", line 448, in check
[01:05:10]     check_command(c, cache)
[01:05:10]   File "/checkout/src/etc/htmldocck.py", line 397, in check_command
[01:05:10]     tree = cache.get_tree(c.args[0])
[01:05:10]   File "/checkout/src/etc/htmldocck.py", line 309, in get_tree
[01:05:10]     raise RuntimeError('Cannot parse an HTML file {!r}: {}'.format(path, e))
[01:05:10] RuntimeError: Cannot parse an HTML file 'foo/fn.f.html': pop from empty stack
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: Arf, forgot to write the condition... (it's not supposed to be there everytime)

**Comment by QuietMisdreavus** [comment]
@bors r+

Thanks! I've wanted a menu like this for a while myself. :D

**Comment by GuillaumeGomez** [comment]
Rebased.

**Pull request merged**